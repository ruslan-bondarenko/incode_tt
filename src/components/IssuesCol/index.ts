export {default as IssuesCol} from './IssuesCol';