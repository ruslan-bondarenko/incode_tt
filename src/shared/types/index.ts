export type ColumnType = {
 id: string;
 title: string;
 issues: IIssue[];
}

export type IUser = {
 login: string;
 id: number;
 node_id: string;
 avatar_url: string;
 gravatar_id: string;
 url: string;
 html_url: string;
 followers_url: string;
 following_url: string;
 gists_url: string;
 starred_url: string;
 subscriptions_url: string;
 organizations_url: string;
 repos_url: string;
 events_url: string;
 received_events_url: string;
 type: string;
 site_admin: boolean;
};

export type ILabels = {
 id: number;
 node_id: string;
 url: string;
 name: string;
 color: string;
 default: boolean;
 description: string | null;
}

export type IReactions = {
 url: string;
 total_count: number;
 '+1': number;
 '-1': number;
 laugh: number;
 hooray: number;
 confused: number;
 heart: number;
 rocket: number;
 eyes: number;
}

export type IIssue = {
 url: string;
 repository_url: string;
 labels_url: string;
 comments_url: string;
 events_url: string;
 html_url: string;
 id: number;
 node_id: string;
 number: number;
 title: string;
 user: IUser;
 labels: ILabels[];
 state: string;
 locked: boolean;
 assignee: any;
 assignees: any[];
 milestone: any;
 comments: number;
 created_at: string;
 updated_at: string;
 closed_at: string | null;
 author_association: string;
 active_lock_reason: string | null;
 body: string;
 reactions: IReactions;
 timeline_url: string;
 performed_via_github_app: any;
 state_reason: string | null;
}

export type IFilteredIssues = { 
 to_do: IIssue[];
 in_progress: IIssue[];
 done: IIssue[];
};

export type IRepoInfo = {
 name: string,
 repo: {
   name: string,
   url: string,
 }
};
