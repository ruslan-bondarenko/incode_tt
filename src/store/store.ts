import {combineReducers, configureStore} from '@reduxjs/toolkit';
import {TypedUseSelectorHook, useDispatch, useSelector} from 'react-redux';

import {issuesSlice} from './slices';

const rootReducer = combineReducers({
  issues: issuesSlice,
})

const initialValue = {
  issues: {
    data: [],
    isLoading: false,
    error: undefined,
    filteredIssues: {
     to_do: [],
     in_progress: [],
     done: [],
    },
    prevUrl: null,
    repoInfo: null
   }
}

export const store = configureStore({
  reducer: {
    issues: issuesSlice,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware()
});

export const createReduxStore = (initialState: RootState = initialValue) => {
  return configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) => getDefaultMiddleware(),
    preloadedState: initialState
  });
}

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

