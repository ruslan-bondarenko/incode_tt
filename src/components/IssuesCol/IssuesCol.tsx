import { IIssue, capitalizeWords } from "@/shared";
import React, { FC } from "react";
import { IssueCard } from "../IssueCard";
import { Col, Typography } from "antd";
import { Droppable } from "react-beautiful-dnd";

type Props = {
  data: IIssue[];
  colKey: string;
};

const IssuesCol: FC<Props> = ({ data, colKey }) => {
  return (
    <Col span={8}>
      <Typography.Title
        level={3}
        style={{ textAlign: "center", margin: "0 0 1rem" }}
      >
        {capitalizeWords(colKey.split("_").join(" "))}
      </Typography.Title>
      <Droppable droppableId={`${colKey}`}>
        {(provided) => (
          <div
            style={{
              backgroundColor: "lightgrey",
              height: "100%",
              minHeight: "50vh",
              padding: "1rem",
            }}
            className="issue-column"
            ref={provided.innerRef}
            {...provided.droppableProps}
          >
            {data.map((issue, index) => (
              <IssueCard
                issue={issue}
                index={index}
                key={`${issue.id}-${index}`}
              />
            ))}
            {provided.placeholder}
          </div>
        )}
      </Droppable>
    </Col>
  );
};

export default IssuesCol;
