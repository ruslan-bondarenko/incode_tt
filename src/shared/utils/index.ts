import { IFilteredIssues, IIssue } from "../types";

export const urlRegex = /https:\/\/github\.com\/[a-zA-Z0-9_.-]+\/[a-zA-Z0-9_.-]+/;

export function capitalizeWords(sentence: string) {
  return sentence.split(' ').map(word => word.charAt(0).toUpperCase() + word.slice(1)).join(' ');
}

export function formatUpdateDate(updateDate: string) {
  const currentDate = new Date();
  const updatedDate = new Date(updateDate);
  const diffTime = Math.abs(currentDate.getTime() - updatedDate.getTime());
  const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
  
  return `updated ${diffDays} days ago`;
}

export const groupIssues = (data: IIssue[]) => {
 const now = new Date();
 const groupedIssues: IFilteredIssues = {
   to_do: [],
   in_progress: [],
   done: [],
 };

 data.forEach((issue: IIssue) => {
   const createdAt = new Date(issue.created_at);
   const isWithin24Hours =
     (now.getTime() - createdAt.getTime()) / (1000 * 60 * 60) <= 24;

   if (issue.state !== "closed" && isWithin24Hours) {
     groupedIssues.to_do.push(issue);
   } else if (issue.state === "open") {
     // && issue.assignees.length
     groupedIssues.in_progress.push(issue);
   } else if (issue.state === "closed") {
     groupedIssues.done.push(issue);
   }
 });

 return groupedIssues;
};
