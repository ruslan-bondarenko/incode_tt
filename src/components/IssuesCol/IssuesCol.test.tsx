import { render, screen } from "@testing-library/react";
import React from "react";

import IssuesCol from "./IssuesCol";
import { describe, it } from "@jest/globals";
import { capitalizeWords } from "@/shared";
import { Provider } from "react-redux";
import * as reduxHooks from "react-redux";
import { createReduxStore, store } from "@/store";
import "react-beautiful-dnd";

declare const expect: jest.Expect;

jest.mock("react-redux");
jest.mock("react-beautiful-dnd");

const mockedData = {
  data: [],
  colKey: "test",
};
describe("IssuesCol component", () => {
  it("IssuesCol renders", () => {
    const mockDispatch = jest.fn();
    jest.spyOn(reduxHooks, "useDispatch").mockReturnValue(mockDispatch);
    jest.spyOn(reduxHooks, "useSelector").mockReturnValue(createReduxStore());

    render(
      <Provider store={store}>
        <IssuesCol data={mockedData.data} colKey={mockedData.colKey} />
      </Provider>
    );

    setTimeout(() => {
      expect(
        screen.getByText(
          capitalizeWords(mockedData.colKey.split("_").join(" "))
        )
      ).toBeInTheDocument();
    }, 0);
  });

  it("IssuesCol snapshot", () => {
    const view = render(
      <IssuesCol data={mockedData.data} colKey={mockedData.colKey} />
    );

    expect(view).toMatchSnapshot();
  });
});
