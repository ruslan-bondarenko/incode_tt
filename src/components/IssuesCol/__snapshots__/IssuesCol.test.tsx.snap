// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssuesCol component IssuesCol snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ant-col ant-col-8 css-dev-only-do-not-override-1kuana8"
      >
        <h3
          class="ant-typography css-dev-only-do-not-override-1kuana8"
          style="text-align: center; margin: 0px 0px 1rem;"
        >
          Test
        </h3>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ant-col ant-col-8 css-dev-only-do-not-override-1kuana8"
    >
      <h3
        class="ant-typography css-dev-only-do-not-override-1kuana8"
        style="text-align: center; margin: 0px 0px 1rem;"
      >
        Test
      </h3>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
